pkg/controller/machinemanagement/machinemanagement_controller.go

```diff
@@ -0,0 +1,316 @@
+		if cd.Spec.MachineManagement.TargetNamespace == "" {
+			cd.Spec.MachineManagement.TargetNamespace = apihelpers.GetResourceName(cd.Name+"-targetns", utilrand.String(5))
+
+			// Ensure the cluster deployment has a finalizer for cleanup
+			if !controllerutil.ContainsFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace) {
+				controllerutil.AddFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace)
+			}
+
+			if err := r.Update(context.TODO(), cd); err != nil {
+				cdLog.WithError(err).Log(controllerutils.LogLevel(err), "failed to update cluster deployment")
+				return reconcile.Result{Requeue: true}, nil
+			}
+		}
+
+		ns := &corev1.Namespace{}
+		if err := r.Get(context.Background(), types.NamespacedName{Name: cd.Spec.MachineManagement.TargetNamespace}, ns); err != nil && apierrors.IsNotFound(err) {
+			cdLog.Info("Creating the target namespace ", cd.Spec.MachineManagement.TargetNamespace)
+			ns.Name = cd.Spec.MachineManagement.TargetNamespace
+			if err := r.Create(context.TODO(), ns); err != nil && !apierrors.IsAlreadyExists(err) {
+				return reconcile.Result{}, fmt.Errorf("failed to create target namespace %q: %w", cd.Spec.MachineManagement.TargetNamespace, err)
+			}
+		}
+
+		// Ensure targetNamespace has machine management annotation
+		if err := r.addAnnotationToTargetNamespace(cd, cdLog, cd.Spec.MachineManagement.TargetNamespace); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Sync credentials secret to targetNamespace
+		if err := r.createOrUpdateSecretInTargetNamespace(utils.CredentialsSecretName(cd), cd, cdLog); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Sync pull secret to targetNamespace
+		if err := r.createOrUpdateSecretInTargetNamespace(cd.Spec.PullSecretRef.Name, cd, cdLog); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Sync SSH key secret to targetNamespace
+		if cd.Spec.Provisioning != nil && cd.Spec.Provisioning.SSHPrivateKeySecretRef != nil {
+			if err := r.createOrUpdateSecretInTargetNamespace(cd.Spec.Provisioning.SSHPrivateKeySecretRef.Name, cd, cdLog); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+// createOrUpdateSecretInTargetNamespace
+func (r *ReconcileMachineManagement) createOrUpdateSecretInTargetNamespace(secretName string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	targetNamespace := cd.Spec.MachineManagement.TargetNamespace
+	secret := &corev1.Secret{}
+	err := r.Get(context.Background(), types.NamespacedName{Name: secretName, Namespace: targetNamespace}, secret)
+	// Create secret in targetNamespace
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cdLog.Infof("Creating secret %s in the target namespace %s", secretName, targetNamespace)
+			if err := r.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: secretName}, secret); err != nil {
+				return fmt.Errorf("failed to get secret %s: %w", secretName, err)
```

Comment:
Suggest `errors.Wrapf()` (throughout)

Reply:
Done throughout the machinemanagement controller.